<a href="https://colab.research.google.com/github/tim-sadler/tutorials/blob/main/CouponAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [37]:
import pandas as pd

# Promotion Analysis: The effect of coupons on customer value

## Why this analysis is relevant

Companies, especially e-commerce business regularly run campaigns to attract customers or boost sales. Coupons are among the most favored promotional instruments, as they have important features:

- they can be sent out only to certain customers, if desired, and
- the customer has to actively use the coupon code in the checkout process, which is why many more customers are exposed to the campaign than end up actually getting the (from the business perspectice) expensive discounts.

However, coupons - really any kind of promotion - comes with a huge disadvantage:

**They might be boosting short-term sales, but they are also likely to attract "discount-prone" customers, that only but when they see a promotion.**

Therefore, the question is, how valuable are the customers we acquire with sich coupons? In this post I show a simple procedure to analyze this question and use the results to improve usage of promotions for customer acquisition.

## Research Question

The RQ behind this analysis is the following: 

*What is the long-term value of customers acquired with coupons vs. organically acquired customers?*

Our a-priori expectation of hypothesis would be: *These customers are less valuable (in terms of revenue generated) than organic customers*.

## Our Example Case

Let's assume we are running this example for the case of an online store for non-food fast-moving consumer goods. let's assume they sell soaps, candles, socks and and other kind of items in the price range of single to small double digit figures.

Our customers can only purchase if they are signed-in or have an email address provided. Like with many e-commerce platform, we can obtain transactional data like this:



In [38]:
transactions = pd.read_csv("transactions_with_coupons.csv", index_col = False)

In [39]:
transactions.head()

customer_id acquisition_date  total_purchases  \
0  00a56493e7dad5c7a0ce38f33fb79a65       2022-01-26                3   
1  00a56493e7dad5c7a0ce38f33fb79a65       2022-01-26                3   
2  00a56493e7dad5c7a0ce38f33fb79a65       2022-01-26                3   
3  00a9a3d7b33d3dcc0b6898d0b342495f       2018-03-03                1   
4  00e246a83cfc888460c786b4664e44b1       2022-02-13               11   

         date  quantity  revenue campaign coupon_code  discount  
0  2022-03-24         5    36.65      NaN         NaN       0.0  
1  2022-08-12         7   100.24      NaN         NaN       0.0  
2  2022-12-08        11   146.52      NaN         NaN       0.0  
3  2020-04-18        10   146.60      NaN         NaN       0.0  
4  2022-03-07         2    17.30      NaN         NaN       0.0

There is one row per customer and transaction with nine columns:

- *customer_id*: unique identifier of the customer
- *acquisition_date*: the date at which the customer had their first transaction (the same for all transactions of the same customer)
- *date*: the date of the transaction
- *revenue*: the total revenue of the transaction (amount paid by the customer)
- *campaign*: name of the campaign the customer redeemed a coupon from
- *coupon_code*: the coupon code used to
obtain the discount
- *discount*: the relative discount obtained with the coupon

The latter three columns are NA, if the transaction did not involve a coupon.

In [40]:
transactions.describe(include = "all", datetime_is_numeric=True)

customer_id acquisition_date  total_purchases  \
count                               2517             2517      2517.000000   
unique                               995             1000              NaN   
top     17544c9a892645cabdfc2a423233801b       2019-02-03              NaN   
freq                                  18               18              NaN   
mean                                 NaN              NaN         4.078268   
std                                  NaN              NaN         2.931354   
min                                  NaN              NaN         1.000000   
25%                                  NaN              NaN         2.000000   
50%                                  NaN              NaN         3.000000   
75%                                  NaN              NaN         5.000000   
max                                  NaN              NaN        18.000000   

              date     quantity      revenue    campaign coupon_code  \
count         2517  2517.000000  2517.000000         161         161   
unique        1055          NaN          NaN           2           2   
top     2022-12-31          NaN          NaN  Newsletter      AXQZEN   
freq            13          NaN          NaN         160         160   
mean           NaN     5.615018    55.875491         NaN         NaN   
std            NaN     2.817018    33.427433         NaN         NaN   
min            NaN     1.000000     5.020000         NaN         NaN   
25%            NaN     4.000000    30.120000         NaN         NaN   
50%            NaN     6.000000    51.310000         NaN         NaN   
75%            NaN     8.000000    76.090000         NaN         NaN   
max            NaN    15.000000   190.500000         NaN         NaN   

           discount  
count   2517.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean       0.006456  
std        0.024881  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        0.250000

Here, we can see that we habe about 1,000 customers making a total of 2,517 transactions.

In [41]:
print(f"Start: {transactions['date'].min()}, End: {transactions['date'].max()}")

Start: 2018-02-03, End: 2022-12-31


The time frame considered is between February 2018 and December 2022.

In [42]:
transactions.groupby("campaign").agg({"discount": "mean", "date": ["min", "max"]})

discount        date            
                   mean         min         max
campaign                                       
Newsletter         0.10  2018-02-03  2022-12-31
Valentines2022     0.25  2022-02-14  2022-02-14

We can also identify two campaigns: Newsletter and Valentines22.

- **Newsletter**: An ongoing campaign yielding ten percent discount for a newsöetter sign-up.
- **Valentines22**: A one-off campaign for Valentine's Day giving a 25 percent discount.

## Data Prep

The above data is a pretty standard format to be obtained from an online store backend. In order to reproduce the following steps of data preparation, make sure you have the data in the same shape, i. e. one row per customer and transacrion / purchase incidence.

### Calculation Discount and Revenue before Discount

The first thing you may habe noticed is the *revenue* column, containing the payment amount for each transaction. We want to determine the *revenue before discount* and the actual *discount* in $.

In [43]:
transactions["discount_amount"] = (transactions["revenue"]*(1+transactions["discount"]))-transactions["revenue"]

In [44]:
transactions["revenue_before_discount"] = transactions["revenue"] + transactions["discount_amount"]

### Splitting between first and later trips

Recall, that we want to know which effect a coupon upon acquisition has on the value of the customer. The contribution to the total value from the first transaction is, due to the discount, systematically biased (i. e., lower) if a coupon is used. Therefore, we will split the transactions into the first and later transactions.

For that we first need to find the rolling number of each transaction.

In [45]:
transactions["transaction_num"] = transactions.groupby("customer_id").cumcount()+1

Lets see how many customer per total number of transactions we have:

In [46]:
transactions.groupby("transaction_num").size()

transaction_num
1     995
2     599
3     357
4     223
5     135
6      78
7      50
8      30
9      19
10      9
11      6
12      3
13      3
14      3
15      3
16      2
17      1
18      1
dtype: int64

In the second row we can see that we have 599 repeating customers with at least two transactions and only 396 customers 

We will now extract the characteristicsof each of the customers' first transaction and append it to a summarized data frame of later transactions.

In [47]:
first_transactions = transactions[transactions["transaction_num"] == 1][["customer_id", "acquisition_date", "quantity", "campaign"]].set_index("customer_id")

Let's do some renaming to make the information easily interpretable and also replace the NAs in campaigns with "None", so they don't get accidentally dropped:

In [48]:
first_transactions.columns = ["acquisition_date", "quantity_first", "campaign_first"]

In [49]:
first_transactions["campaign_first"] = first_transactions["campaign_first"].fillna("None")

In [50]:
later_transactions = transactions[transactions["transaction_num"] > 1].groupby("customer_id").agg({"customer_id": "count", "revenue": "sum", "quantity": "mean", "date": "max", "discount": "mean"})

We are aggregating the later transactions for each customer and get the total number of later transactions, total revenue (i. e., value), the mean quantity purchased, the date of latest purchase and the mean discount from later transactions. We elaborate on the need for each of these variables later. First, lets do some
renaming for better interpretability:

In [51]:
later_transactions.columns = ["transactions_num_later", "value", "quantity_later", "max_date", "discounts_later"]

Now, let's merge the aggregated later transactions with the characteristics of the first transactions.

In [52]:
later_transactions = later_transactions.merge(first_transactions, left_index = True, right_index = True, how = "left")

We now have a data frame with one row per customer and the information of later and first transactions as columns. Let's take a brief look:

In [53]:
later_transactions.groupby("campaign_first")[["value", "transactions_num_later"]].mean()

value  transactions_num_later
campaign_first                                    
Newsletter       78.673226                1.935484
None            145.393979                2.573944

At first glance it looks like, customers acquired with a *Newletter* coupon indeed create less value in later transactions and the column *transactions_num_later* indicates that they do so due to less frequent transactions.

**However**: There might be a bias here. It might happen that *organic* customers have longer tenure or retention (i. e. joined earlier) than customers acquired with the *Newsletter* coupon. We thus have to correct for potential differences in the length of the customer lifetime duration. 

### Correcting for Different Lifetime Lengths

There are different options to do so, the simplest being to calculate the duration between the acquisition date and the date of the last transaction and divide the value by it. However, this opens up the possibility of another bias: customers who joined very late did not have the ability to accumulate a long lifetime, still their short lifetime would get weighted more strongly when dividing by a smaller number.

The second option would be to remove, e. g. the most recent year from the data and thua remove the bias created by more recent customers. This however would also reduce the number of customers to analyze.

We therefore chose a third option: divide the value (and # of transactions) by the remaining time left in the data to observe the customer. 

Example: a customer males their most recent transaction in the last week we observe. Wenthen divide their total value by 1 (week left), because we now, they are still an active customer. Correspondingly, a customer who has not made a transaction for 50 weeks has a much lower corrected value, as their chance of still being active is much lower.

In [54]:
later_transactions["remaining_weeks"] = (pd.to_datetime(later_transactions["max_date"].max()) - pd.to_datetime(later_transactions["max_date"])).dt.days//7 +1

In [59]:
later_transactions["value_corrected"] = later_transactions['value']/later_transactions["remaining_weeks"]
later_transactions["transactions_num_later_corrected"] = later_transactions["transactions_num_later"]/later_transactions["remaining_weeks"]

Above we calculated the remaining weeks in the data per customer (the minimum being one week) and divided the variables of interest by that value. We indicate the new values by the suffix *_corrected*.

In [60]:
later_transactions.groupby("campaign_first")[["value_corrected", "transactions_num_later_corrected"]].mean()

value_corrected  transactions_num_later_corrected
campaign_first                                                   
Newsletter            16.719509                          0.345715
None                  24.921393                          0.447436

Displaying the mean corrected value and mean corrected later transactions, we see that also after correcting for the bias, 

# Neuer Abschnitt